File: src/Shortest Path in 1-2 Graph/Code/solution.java
import java.util.*;

class Solution {
    public int shortestPath(int V, int src, int dest, int[][] edges) {

        // Total nodes after adding virtual nodes
        int totalNodes = V;

        // Adjacency list for transformed graph
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();

        // Maximum possible nodes = V + number of edges
        for (int i = 0; i < V + edges.length; i++) {
            adj.add(new ArrayList<>());
        }

        // Build transformed graph
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];

            if (w == 1) {
                // Normal edge
                adj.get(u).add(v);
                adj.get(v).add(u);
            } else {
                // Create virtual node
                int newNode = totalNodes++;

                // Replace weight 2 edge
                adj.get(u).add(newNode);
                adj.get(newNode).add(u);

                adj.get(newNode).add(v);
                adj.get(v).add(newNode);
            }
        }

        // Distance array
        int[] dist = new int[totalNodes];
        Arrays.fill(dist, -1);

        Queue<Integer> queue = new LinkedList<>();

        // Start BFS
        queue.offer(src);
        dist[src] = 0;

        while (!queue.isEmpty()) {
            int node = queue.poll();

            for (int next : adj.get(node)) {
                if (dist[next] == -1) {
                    dist[next] = dist[node] + 1;
                    queue.offer(next);
                }
            }
        }

        // Return answer
        return dist[dest];
    }
}

File: src/Shortest Path in 1-2 Graph/Code/solution.py
from collections import deque

class Solution:
    def shortestPath(self, V: int, src: int, dest: int, edges: list[list[int]]) -> int:

        # Total nodes after adding virtual nodes
        total_nodes = V

        # Adjacency list for transformed graph
        adj = [[] for _ in range(V + len(edges))]

        # Build transformed graph
        for u, v, w in edges:

            if w == 1:
                # Normal edge
                adj[u].append(v)
                adj[v].append(u)
            else:
                # Create one virtual node
                new_node = total_nodes
                total_nodes += 1

                # Replace weight 2 edge with two weight 1 edges
                adj[u].append(new_node)
                adj[new_node].append(u)

                adj[new_node].append(v)
                adj[v].append(new_node)

        # Distance array
        dist = [-1] * total_nodes

        # BFS queue
        q = deque()

        q.append(src)
        dist[src] = 0

        while q:

            node = q.popleft()

            for nxt in adj[node]:

                if dist[nxt] == -1:
                    dist[nxt] = dist[node] + 1
                    q.append(nxt)

        # Return shortest distance to destination
        return dist[dest]